<a href="https://colab.research.google.com/github/ev-greshkina/DataAnalyst-educational-course/blob/main/%22%D0%98%D1%82%D0%BE%D0%B3%D0%BE%D0%B2%D0%B0%D1%8F_%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0_%D0%90%D0%B2%D0%B8%D0%B0%D1%81%D0%BE%D0%BE%D0%B1%D1%89%D0%B5%D0%BD%D0%B8%D0%B5_%D0%A1%D0%A8%D0%90_v_2_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Шаг 1: Загрузка и предобработка данных

In [1]:
import pandas as pd

# Загрузка данных
flights = pd.read_csv('flights.csv', dtype={'ORIGIN_AIRPORT':str, 'DESTINATION_AIRPORT':str})
airports = pd.read_csv('airports.csv')
airlines = pd.read_csv('airlines.csv')

**Задача 1: Найти аэропорт с минимальной задержкой вылета**

In [32]:
# Создать копию DataFrame 'flights' для выполнения преобразований и вычислений с копией
# а не с оригиналом
flights_task_1 = flights.copy()

In [33]:
# Преобразование задержки в числовой формат
flights_task_1['DEPARTURE_DELAY'] = pd.to_numeric(flights_task_1['DEPARTURE_DELAY'], errors='coerce')

# Удаляем строки с пустыми значениями задержек
flights_task_1 = flights_task_1.dropna(subset=['DEPARTURE_DELAY', 'ORIGIN_AIRPORT'])

flights_task_1

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

In [34]:
# В столбце ORIGIN_AIRPORT имеются цифровые ITA-CODE. Цифровые обозначения
# вероятно являются ошибочно внесенными данныи, поэтому строки содержащие цифры в
# столбцах ORIGIN_AIRPORT и DESTINATION_AIRPORT подлежат удалению. Идентифицировать аэропрт
# по этим данным не возможно
# Удаляем строки, содержащие цифры

# Преобразовать данные в столбце ORIGIN_AIRPORT в строковые значения
flights_task_1['ORIGIN_AIRPORT'] = flights_task_1['ORIGIN_AIRPORT'].astype(str)

# Фильтрация строк, содержащих цифры
flights_task_1 = flights_task_1[~flights_task_1['ORIGIN_AIRPORT'].str.contains('[0-9]')]
flights_task_1

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

In [35]:
# Группировка по аэропорту и расчет средней задержки вылета
airport_delay = flights_task_1.groupby('ORIGIN_AIRPORT')['DEPARTURE_DELAY'].mean().reset_index()
airport_delay

ORIGIN_AIRPORT  DEPARTURE_DELAY
0              ABE         8.844812
1              ABI         5.783989
2              ABQ         8.877015
3              ABR        10.101056
4              ABY         8.880046
..             ...              ...
317            WRG         3.091884
318            WYS         1.375000
319            XNA        11.886184
320            YAK        -6.292237
321            YUM         1.232470

[322 rows x 2 columns]

In [36]:
# Таблица полученных значений содержит отрицательные значения, которые вероятно
# означают, что самолет вылетел с опережением графика. Для получения данных о наименьшей
# задержке вылета требуется удалить все строки с отрицательным значением DEPARTURE DELAY
airport_delay = airport_delay.drop(airport_delay.index[airport_delay['DEPARTURE_DELAY'] < 0])

# Отсортировать и вывести полученные значения
sorted = airport_delay.sort_values(by='DEPARTURE_DELAY')
sorted

ORIGIN_AIRPORT  DEPARTURE_DELAY
274            SCC         0.154122
10             ADQ         0.405034
184            LIH         0.556254
89             DLG         0.558442
46             BTM         0.674383
..             ...              ...
231            OTH        17.777358
293            STC        18.692308
147            HYA        23.182927
219            MVY        25.907317
154            ILG        29.391753

[311 rows x 2 columns]

In [37]:
# Поиск аэропорта с минимальной задержкой
# Если значение задержки вылета отрицательное, то это означает, что вылет выполнен с опережением графика
min_delay_airport = airport_delay.loc[airport_delay['DEPARTURE_DELAY'].idxmin()]
min_delay_airport

ORIGIN_AIRPORT          SCC
DEPARTURE_DELAY    0.154122
Name: 274, dtype: object

In [38]:
# Получение информации о названии аэропорта
airport_info = airports[airports['IATA_CODE'] == min_delay_airport['ORIGIN_AIRPORT']]
airport_info

IATA_CODE                                  AIRPORT       CITY STATE  \
274       SCC  Deadhorse Airport (Prudhoe Bay Airport)  Deadhorse    AK   

    COUNTRY  LATITUDE  LONGITUDE  
274     USA  70.19476 -148.46516

In [39]:
# Вывод результата
print(f"Минимальная средняя задержка вылета: {airport_info['AIRPORT'].values[0]} ({min_delay_airport['ORIGIN_AIRPORT']}) с задержкой {min_delay_airport['DEPARTURE_DELAY']:.2f} минут")

Минимальная средняя задержка вылета: Deadhorse Airport (Prudhoe Bay Airport) (SCC) с задержкой 0.15 минут


**Задача 2: Самая пунктуальная авиакомпания на прилет в Los Angeles International Airport**

In [40]:
# Создать копию DataFrame 'flights' для выполнения преобразований и вычислений с копией
# а не с оригиналом
flights_task_2 = flights.copy()

In [41]:
# Преобразование задержки в числовой формат
flights_task_2['ARRIVAL_DELAY'] = pd.to_numeric(flights_task_2['ARRIVAL_DELAY'], errors='coerce')


In [42]:
# Удаление строк с пустыми значениями в столбцах 'ARRIVAL_DELAY' и 'DESTINATION_AIRPORT'
flights_task_2 = flights_task_2.dropna(subset=['ARRIVAL_DELAY', 'DESTINATION_AIRPORT'])

In [43]:
# Группировка по авиакомпаниям и расчет средней задержки для рейсов в Лос-Анджелес
la_flights = flights_task_2[flights_task_2['DESTINATION_AIRPORT'] == 'LAX']

In [44]:
# Проверка, есть ли данные для анализа
if not la_flights.empty:
    airline_punctuality = la_flights.groupby('AIRLINE')['ARRIVAL_DELAY'].mean().reset_index()

    # Поиск самой пунктуальной авиакомпании
    if not airline_punctuality.empty:
        most_punctual_airline = airline_punctuality.loc[airline_punctuality['ARRIVAL_DELAY'].idxmin()]

        # Получение информации о названии авиакомпании
        airline_info = airlines[airlines['IATA_CODE'] == most_punctual_airline['AIRLINE']]

        if not airline_info.empty:
            print(f"Самая пунктуальная авиакомпания на прилет в LAX: {airline_info['AIRLINE'].values[0]} с задержкой {most_punctual_airline['ARRIVAL_DELAY']:.2f} минут")
        else:
            print("Информация об авиакомпании не найдена.")
    else:
        print("Нет данных для анализа пунктуальности авиакомпаний.")
else:
    print("Нет рейсов, прилетающих в LAX для анализа.")

Самая пунктуальная авиакомпания на прилет в LAX: Hawaiian Airlines Inc. с задержкой -4.67 минут


**Задача 3: Найти аэропорт, где самолёты проводят больше всего времени на рулении (среднее значение) (переработано)**

In [2]:
# Создать копию DataFrame 'flights' для выполнения преобразований и вычислений с копией
# а не с оригиналом
flights_task_3 = flights.copy()

In [3]:
# Преобразование времени руления в числовой формат
flights_task_3['TAXI_IN'] = pd.to_numeric(flights_task_3['TAXI_IN'], errors='coerce')
flights_task_3['TAXI_OUT'] = pd.to_numeric(flights_task_3['TAXI_OUT'], errors='coerce')

In [4]:
# Удаляем строки с пустыми значениями столбца ORIGIN_AIRPORT
flights_task_3 = flights_task_3.dropna(subset=['ORIGIN_AIRPORT'])
flights_task_3

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

In [5]:
# В столбце ORIGIN_AIRPORT имеются цифровые ITA-CODE. Цифровые обозначения
# вероятно являются ошибочно внесенными данныи, поэтому строки содержащие цифры в
# столбцах ORIGIN_AIRPORT и DESTINATION_AIRPORT подлежат удалению. Идентифицировать аэропрт
# по этим данным не возможно
# Удаляем строки, содержащие цифры

# Преобразовать данные в столбце ORIGIN_AIRPORT в строковые значения
flights_task_3['ORIGIN_AIRPORT'] = flights_task_3['ORIGIN_AIRPORT'].astype(str)

# Фильтрация строк, содержащих цифры
flights_task_3 = flights_task_3[~flights_task_3['ORIGIN_AIRPORT'].str.contains('[0-9]')]

flights_task_3

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

In [6]:
# Вычисление среднего времени руления
taxi_time = flights_task_3.groupby('ORIGIN_AIRPORT')[['TAXI_OUT', 'TAXI_IN']].mean().reset_index()

taxi_time

ORIGIN_AIRPORT   TAXI_OUT    TAXI_IN
0              ABE  13.408949   8.460161
1              ABI   9.188004  10.685932
2              ABQ  11.987672   7.670164
3              ABR  16.600302   7.717949
4              ABY  12.084198   7.794694
..             ...        ...        ...
317            WRG   7.820827   4.178737
318            WYS  11.903846   5.975962
319            XNA  14.681277  11.446472
320            YAK   7.919330   3.935385
321            YUM  14.348436  10.468716

[322 rows x 3 columns]

In [7]:
# Расчет общего среднего времени руления
taxi_time['AVG_TAXI_TIME'] = taxi_time['TAXI_OUT'] + taxi_time['TAXI_IN']

taxi_time['AVG_TAXI_TIME']

0      21.869110
1      19.873935
2      19.657836
3      24.318250
4      19.878893
         ...    
317    11.999563
318    17.879808
319    26.127749
320    11.854715
321    24.817152
Name: AVG_TAXI_TIME, Length: 322, dtype: float64

In [8]:
# Нахождение аэропорта с максимальным временем руления
max_taxi_airport = taxi_time.loc[taxi_time['AVG_TAXI_TIME'].idxmax()]

max_taxi_airport

ORIGIN_AIRPORT          LGA
TAXI_OUT           26.72137
TAXI_IN            7.818918
AVG_TAXI_TIME     34.540288
Name: 182, dtype: object

In [9]:
# Получение информации о аэропорту
max_taxi_airport_info = airports[airports['IATA_CODE'] == max_taxi_airport['ORIGIN_AIRPORT']]

max_taxi_airport_info

IATA_CODE                                  AIRPORT      CITY STATE  \
182       LGA  LaGuardia Airport (Marine Air Terminal)  New York    NY   

    COUNTRY  LATITUDE  LONGITUDE  
182     USA  40.77724  -73.87261

In [10]:
# Вывод результата
print(f"Аэропорт с максимальным средним временем руления: {max_taxi_airport_info['AIRPORT'].values[0]} ({max_taxi_airport['ORIGIN_AIRPORT']}) - Среднее время руления: {max_taxi_airport['AVG_TAXI_TIME']:.2f} минут")

Аэропорт с максимальным средним временем руления: LaGuardia Airport (Marine Air Terminal) (LGA) - Среднее время руления: 34.54 минут


**Задача 4: Построить модель для выбора топ-3 аэропортов прилета (переработано)**

In [28]:
# Подключение дополнительных библиотек
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_squared_error as mse

In [12]:
# Преобразование задержки в числовой формат с оптимизацией памяти
flights['DEPARTURE_DELAY'] = pd.to_numeric(flights['DEPARTURE_DELAY'], errors='coerce', downcast='float')
flights['ARRIVAL_DELAY'] = pd.to_numeric(flights['ARRIVAL_DELAY'], errors='coerce', downcast='float')

In [13]:
# Убираем строки с отсутствующими значениями
flights = flights.dropna(subset=['DEPARTURE_DELAY', 'ARRIVAL_DELAY', 'ORIGIN_AIRPORT', 'DESTINATION_AIRPORT'])

In [14]:
# Преобразовать данные в столбце ORIGIN_AIRPORT в строковые значения
flights['ORIGIN_AIRPORT'] = flights['ORIGIN_AIRPORT'].astype(str)

In [15]:
# Фильтрация строк, содержащих цифры
flights = flights[~flights['ORIGIN_AIRPORT'].str.contains('[0-9]')]

In [16]:
# Выполнить замену категориальных данных в столбце ORIGIN_AIRPORT
# Извлеч столбец ORIGIN_AIRPORT в отдельный Series
airport = flights.ORIGIN_AIRPORT

# Удалить дубликаты названий аэропортов
airport = airport.drop_duplicates()

# Сбросить параметры индексирования методом reset_index
airport = airport.reset_index()

# Удалить стобец index для т.к. сохранены исходыне индексы и они не по порядку.
# Сохранить только столбец ORIGIN_AIRPORT
airport = airport.ORIGIN_AIRPORT

# Сбросить значение index
airport = airport.reset_index()
airport

index ORIGIN_AIRPORT
0        0            ANC
1        1            LAX
2        2            SFO
3        3            SEA
4        4            LAS
..     ...            ...
317    317            WYS
318    318            DLG
319    319            AKN
320    320            GST
321    321            HYA

[322 rows x 2 columns]

In [17]:
# Произвести замену значения в столбце ORIGIN_AIRPORT дата фрейма flights на соответствующее значение
# из созданной таблицы числовых замен значения аэропорта
# Объединить таблицы методом merge
flights = flights.merge(airport, how='left', on='ORIGIN_AIRPORT', suffixes=('', '1'))
flights

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5231125  2015     12   31            4      B6            688      N657JB   
5231126  2015     12   31            4      B6            745      N828JB   
5231127  2015     12   31            4      B6           1503      N913JB   
5231128  2015     12   31            4      B6            333      N527JB   
5231129  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5231125            LAX                 BOS                 2359  ...   
5231126            JFK                 PSE                 2359  ...   
5231127            JFK                 SJU                 2359  ...   
5231128            MCO                 SJU                 2359  ...   
5231129            JFK                 BQN                 2359  ...   

         ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  \
0                -22.0         0          0                  NaN   
1                 -9.0         0          0                  NaN   
2                  5.0         0          0                  NaN   
3                 -9.0         0          0                  NaN   
4                -21.0         0          0                  NaN   
...                ...       ...        ...                  ...   
5231125          -26.0         0          0                  NaN   
5231126          -16.0         0          0                  NaN   
5231127           -8.0         0          0                  NaN   
5231128          -10.0         0          0                  NaN   
5231129            2.0         0          0                  NaN   

         AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
0                     NaN             NaN            NaN                  NaN   
1                     NaN             NaN            NaN                  NaN   
2                     NaN             NaN            NaN                  NaN   
3                     NaN             NaN            NaN                  NaN   
4                     NaN             NaN            NaN                  NaN   
...                   ...             ...            ...                  ...   
5231125               NaN             NaN            NaN                  NaN   
5231126               NaN             NaN            NaN                  NaN   
5231127               NaN             NaN            NaN                  NaN   
5231128               NaN             NaN            NaN                  NaN   
5231129               NaN             NaN            NaN                  NaN   

         WEATHER_DELAY  index  
0                  NaN      0  
1                  NaN      1  
2                  NaN      2  
3                  NaN      1  
4                  NaN      3  
...                ...    ...  
5231125            NaN      1  
5231126            NaN     38  
5231127            NaN     38  
5231128            NaN     20  
5231129            NaN     38  

[5231130 rows x 32 columns]

In [19]:
# Переименовать столбец ORIGIN_AIRPORT
flights = flights.rename(columns={'ORIGIN_AIRPORT': 'RENAMED_ORIGIN_AIRPORT'})

# Переименовать столбец index в ORIGIN_AIRPORT
flights = flights.rename(columns={'index': 'ORIGIN_AIRPORT'})

flights

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5231125  2015     12   31            4      B6            688      N657JB   
5231126  2015     12   31            4      B6            745      N828JB   
5231127  2015     12   31            4      B6           1503      N913JB   
5231128  2015     12   31            4      B6            333      N527JB   
5231129  2015     12   31            4      B6            839      N534JB   

        RENAMED_ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                          ANC                 SEA                    5  ...   
1                          LAX                 PBI                   10  ...   
2                          SFO                 CLT                   20  ...   
3                          LAX                 MIA                   20  ...   
4                          SEA                 ANC                   25  ...   
...                        ...                 ...                  ...  ...   
5231125                    LAX                 BOS                 2359  ...   
5231126                    JFK                 PSE                 2359  ...   
5231127                    JFK                 SJU                 2359  ...   
5231128                    MCO                 SJU                 2359  ...   
5231129                    JFK                 BQN                 2359  ...   

         ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  \
0                -22.0         0          0                  NaN   
1                 -9.0         0          0                  NaN   
2                  5.0         0          0                  NaN   
3                 -9.0         0          0                  NaN   
4                -21.0         0          0                  NaN   
...                ...       ...        ...                  ...   
5231125          -26.0         0          0                  NaN   
5231126          -16.0         0          0                  NaN   
5231127           -8.0         0          0                  NaN   
5231128          -10.0         0          0                  NaN   
5231129            2.0         0          0                  NaN   

         AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
0                     NaN             NaN            NaN                  NaN   
1                     NaN             NaN            NaN                  NaN   
2                     NaN             NaN            NaN                  NaN   
3                     NaN             NaN            NaN                  NaN   
4                     NaN             NaN            NaN                  NaN   
...                   ...             ...            ...                  ...   
5231125               NaN             NaN            NaN                  NaN   
5231126               NaN             NaN            NaN                  NaN   
5231127               NaN             NaN            NaN                  NaN   
5231128               NaN             NaN            NaN                  NaN   
5231129               NaN             NaN            NaN                  NaN   

         WEATHER_DELAY  ORIGIN_AIRPORT  
0                  NaN               0  
1                  NaN               1  
2                  NaN               2  
3                  NaN               1  
4                  NaN               3  
...                ...             ...  
5231125            NaN               1  
52311

In [20]:
# Создание исходных данных для построения модели
X = flights[['DEPARTURE_DELAY', 'ORIGIN_AIRPORT']]
y = flights['ARRIVAL_DELAY']

In [21]:
# Разделение на обучающую и тестовую выборки
%%time
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

CPU times: user 1 s, sys: 6.69 ms, total: 1.01 s
Wall time: 1.05 s


In [22]:
# Построение модели
%%time
model = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
model.fit(X_train, y_train)

CPU times: user 20min 1s, sys: 3.82 s, total: 20min 5s
Wall time: 12min 35s


RandomForestRegressor(n_jobs=-1, random_state=42)

In [23]:
# Прогнозирование
y_pred = model.predict(X_test)

In [29]:
# Вычисляем RMSE
rmse = np.sqrt(mse(y_test, y_pred))
print(f"RMSE модели: {rmse:.2f}")

RMSE модели: 12.96


In [30]:
# Прогнозирование задержки по всем аэропортам
all_predictions = model.predict(X)
flights['Predicted_ARRIVAL_DELAY'] = all_predictions

In [40]:
# Случайный выбор аэропорта вылета
#value = df.loc[df['another_column'] == 'value', 'column_name']
random_origin = flights['ORIGIN_AIRPORT'].sample(1).values[0]
airport_name = flights.loc[flights['ORIGIN_AIRPORT'] == random_origin, 'RENAMED_ORIGIN_AIRPORT'].iloc[0]
print(f"Случайно выбранный аэропорт вылета: {airport_name}")

Случайно выбранный аэропорт вылета: SEA


In [51]:
# Получить данные о наименовании аэропорта
airport_info = airports[airports['IATA_CODE'] == airport_name]
airport_info = airport_info.reset_index()
airport_info

index IATA_CODE                               AIRPORT     CITY STATE  \
0    277       SEA  Seattle-Tacoma International Airport  Seattle    WA   

  COUNTRY  LATITUDE  LONGITUDE  
0     USA  47.44898 -122.30931

In [41]:
# Получение возможных направлений из выбранного аэропорта
possible_destinations = flights[flights['ORIGIN_AIRPORT'] == random_origin]['DESTINATION_AIRPORT'].unique()

In [42]:
# Оценка вероятности опоздания для каждого направления
destination_predictions = flights[flights['ORIGIN_AIRPORT'] == random_origin].groupby('DESTINATION_AIRPORT')['Predicted_ARRIVAL_DELAY'].mean().reset_index()

In [43]:
# Добавление информации о направлениях
destination_info = airports[airports['IATA_CODE'].isin(destination_predictions['DESTINATION_AIRPORT'])]

In [44]:
# Объединение данных о задержках с информацией о аэропортах
destination_info = destination_info.merge(destination_predictions, left_on='IATA_CODE', right_on='DESTINATION_AIRPORT', how='left')

In [54]:
# В полученных данных имеются отрицательные значения в столбце Predicted_ARRIVAL_DELAY, что
# вероятно означает прибытие с опережением. Т.к. поставлена задача определить минимальную задержку, то
# необходимо удалить все строки, в которых значение Predicted_ARRIVAL_DELAY отрицательное
destination_info = destination_info.drop(destination_info.index[destination_info['Predicted_ARRIVAL_DELAY'] < 0])

In [46]:
# Получение топ-3 аэропортов с минимальными задержками
top_3_airports = destination_info.nsmallest(3, 'Predicted_ARRIVAL_DELAY')

In [53]:
# Вывод результатов
print(f"\nВозможные направления из аэропорта {airport_info.AIRPORT[0]} с минимальной вероятной задержкой прибытия:")
for index, row in top_3_airports.iterrows():
    print(f"{row['AIRPORT']} ({row['IATA_CODE']}): Средняя задержка прилета {row['Predicted_ARRIVAL_DELAY']:.2f} минут")


Возможные направления из аэропорта Seattle-Tacoma International Airport с минимальной вероятной задержкой прибытия:
Lihue Airport (LIH): Средняя задержка прилета 0.07 минут
Fairbanks International Airport (FAI): Средняя задержка прилета 0.21 минут
Bellingham International Airport (BLI): Средняя задержка прилета 0.37 минут


# Ответы
**Задача 1: Минимальная средняя задержка вылета: Deadhorse Airport (Prudhoe Bay Airport) (SCC) с задержкой 0.15 минут**
****

**Задача 2: Самая пунктуальная авиакомпания на прилет в LAX: Hawaiian Airlines Inc. с задержкой -4.67 минут**
****

**Задача 3: Аэропорт с максимальным средним временем руления: LaGuardia Airport (Marine Air Terminal) (LGA) - Среднее время руления: 34.54 минут**

****
**Задача 4:**
** **
**При решении аэропорт вылета выбирается случайным образом, поэтому для каждого эксперимента топ-3 аэропортов прибытия с минимальной возможной задержкой будет различным**
*   RMSE модели 12.96

**Возможные направления из аэропорта Seattle-Tacoma International Airport с минимальной вероятной задержкой прибытия:**
* Lihue Airport (LIH): Вероятная задержка прилета 0.07 минут
* Fairbanks International Airport (FAI): Вероятная задержка прилета 0.21 минут
* Bellingham International Airport (BLI): Вероятная задержка прилета 0.37 минут